src/librustc_typeck/check/mod.rs

```diff
@@ -1763,47 +1759,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
             debug!("select_all_obligations_and_apply_defaults: defaults={:?}", default_map);
 
-            // We loop over the unsolved variables, resolving them and if they are
-            // and unconstrainted numberic type we add them to the set of unbound
-            // variables. We do this so we only apply literal fallback to type
-            // variables without defaults.
+            // Examine all unsolved variables, and narrow them to the set that have applicable
+            // defaults. We want to process any unsolved variables that have either an explicit
+            // user default, literal fallback, or are diverging.
             for ty in &unsolved_variables {
                 let resolved = self.infcx().resolve_type_vars_if_possible(ty);
```

Comment:
So, I think that `resolved == ty` must always hold at this point in the code. (I can't see why that would not be the case.) If so, we could simplify this code by removing this call, and removing the various matches below.